### PR TITLE
Refactored Coalesce to use appropriate SqlCoalesceExpression

### PR DESCRIPTION
**Pull request by sdanyliv**
Based on https://github.com/linq2db/linq2db/pull/4533, so it is blocked until parent branch is merged.

### CODE REVIEW

**Comment by MaceWindu** [comment]
OrderByTests.OrderBy7 baselines are incorrect (also looks like test data is not good as it didn't failed test, maybe try another divider?)

**Review by MaceWindu — CHANGES_REQUESTED** [review]
I would prefer orderby builder changes and test renames going to separate PRs (2 more) to simplify baselines review

File: Source/LinqToDB/DataProvider/Oracle/OracleSqlExpressionConvertVisitor.cs [hunk 1]
```diff
@@ -128,9 +128,6 @@ public override ISqlExpression ConvertSqlFunction(SqlFunction func)
 		{
 			switch (func)
 			{
-				case { Name: "Coalesce", Parameters.Length: 2 }:
```
Comment (MaceWindu) [inline Source/LinqToDB/DataProvider/Oracle/OracleSqlExpressionConvertVisitor.cs:131]: 👍

File: Source/LinqToDB/SqlProvider/BasicSqlOptimizer.cs [hunk 2]
```diff
@@ -1866,6 +1866,12 @@ protected IQueryElement OptimizeQueries(IQueryElement startFrom, IQueryElement r
 
 #if DEBUG
 			// ReSharper disable once NotAccessedVariable
+
+			if (startFrom is SqlSelectStatement)
```
Comment (viceroypenguin) [inline Source/LinqToDB/SqlProvider/BasicSqlOptimizer.cs:1870]: Dead code?

File: Source/LinqToDB/Linq/Builder/ExpressionBuilder.SqlBuilder.cs [hunk 3]
```diff
@@ -2350,7 +2334,8 @@ Expression GenerateConstructorComparison(SqlGenericConstructorExpression leftCon
 					}
 				}
 
-				// Force nullability
+				// Force nullability.
+				// TODO: Review. Actually it can be removed. Currently only strange NullableBoolTests.NullTest test fails.
```
Comment (viceroypenguin) [inline Source/LinqToDB/Linq/Builder/ExpressionBuilder.SqlBuilder.cs:2338]: Does this need to be reviewed?

File: Source/LinqToDB/SqlQuery/QueryHelper.cs [hunk 4]
```diff
@@ -229,15 +229,24 @@ static bool IsDependsOn(IQueryElement testedRoot, IQueryElement onElement, HashS
 				{
 					var function = (SqlFunction)expr;
 
-					//TODO: unify function names and put in common constant storage
-					//For example it should be "$COALESCE$" and "$CASE$" do do not mix with user defined extension
-
-					if (function.Name is "Coalesce" or PseudoFunctions.COALESCE && function.Parameters.Length == 2)
+					//TODO: probably remove, we have SqlCoalesceExpression
```
Comment (viceroypenguin) [inline Source/LinqToDB/SqlQuery/QueryHelper.cs:232]: Should this be confirmed?

File: Source/LinqToDB/Linq/Builder/ExpressionBuilder.SqlBuilder.cs (hunk 3 above)
Comment (sdanyliv) [reply Source/LinqToDB/Linq/Builder/ExpressionBuilder.SqlBuilder.cs:2338]: If you can yes. I have no time yet.

File: Source/LinqToDB/SqlQuery/QueryHelper.cs (hunk 4 above)
Comment (sdanyliv) [reply Source/LinqToDB/SqlQuery/QueryHelper.cs:232]: If you can handle it, I have no objections.

**Pull request merged**